First pass at an update to 1.21.2.

gniftygnome opened the pull request:
This PR makes the API generally work from my testing so far.  The boat API is completely replaced under the hood, but the API interfaces are similar to before.  The main change is treating all boat and raft types of a single wood type as a group.  API calls use the group's Identifier and the API becomes the authoritative source of other identifiers.  Maybe some people won't like this; I think it is worth the convenience and simplicity.  One reasonable question is whether TerraformBoatData should be part of the API.

- Make it build against 1.21.2
- Fix broken things
- Replace basically all of the boat API

gniftygnome committed:
First pass at an update to 1.21.2.

- Make it build against 1.21.2
- Fix broken things
- Replace basically all of the boat API

gniftygnome committed:
I forgot about PillarLogHelper...

- PillarLogHelper.of methods removed
- PillarLogHelper.settings methods added
  They provide AbstractBlock.Settings instead of PillarLog

- Improvements to the testmod

gniftygnome committed:
Darn it IDEA, why do you keep removing this change from my commits?

gniftygnome commented:
Things I've noticed but don't yet understand:

* Now that they have a texture, I see the custom rafts do not render quite right.  There is a tiny one texture pixel high space above any edge viewed across the raft where you can see into the water.  I have absolutely no idea what is going on there.
* In just `MixinEatGrassGoal`, IDEA claims `org.spongepowered.asm.mixin.injection.callback.LocalCapture` does not exist (other classes in the same package do exist).  I have tried clearing caches and restarting and so on, but it persists.  However, this seems cosmetic; the build is fine and works fine...

gniftygnome commented:
BTW I have Traverse fully updated to use this Terraform version, so if anybody wants that for testing, let me know and I'll push it.

haykam821 commented on gradle/wrapper/gradle-wrapper.properties:
The Gradle version is updated here; should the other Gradle files be updated as well?

gniftygnome replied:
I don't usually bother, but sure, I can go refresh myself on how to do that.

gniftygnome replied:
Done in  [1448a67](https://github.com/TerraformersMC/Terraform/pull/95/commits/1448a6769f1629dc745202fe9d4bb233b9e90d12).

haykam821 commented on terraform-wood-api-v1/src/main/java/com/terraformersmc/terraform/boat/api/item/TerraformBoatItemHelper.java:
```diff
@@ -1,91 +1,101 @@
+	 *
+	 * This method should be called once for each boat type.
+	 * Created items and entities will have identifiers similar to
+	 * {@code id.withSuffixedPath("_boat")} and {@code id.withPrefixedPath("chest_raft/")}.
```
Would the item ID be `examplemod:mahogany_boat` and the entity type ID be `examplemod:boat/mahogany`, then? I assume the second identifier refers to the entity model layer.

gniftygnome replied:
Yes.  Now that we are doing this entirely the vanilla way, we create an item, entity type, and model layer.  For our mods, there is no need to know the identifiers (although they are predictable, as needed to put the assets in the right place).  I doubt anybody needs the model layer.  I have wondered if others might need either the identifiers or the entity type.  If we think folks might need things like these, I could refactor so the getters of TerraformBoatData are exposed in the API.

haykam821 commented on terraform-wood-api-v1/src/main/java/com/terraformersmc/terraform/wood/api/block/PillarLogHelper.java:
```diff
@@ -12,70 +12,66 @@ private PillarLogHelper() {
 	 */
-	public static PillarBlock ofNether(MapColor wood, MapColor bark) {
-		return new PillarBlock(AbstractBlock.Settings.create()
+	public static AbstractBlock.Settings settingsNether(MapColor wood, MapColor bark) {
```
```suggestion
	public static AbstractBlock.Settings createNetherSettings(MapColor wood, MapColor bark) {
```

haykam821 commented on terraform-wood-api-v1/src/main/java/com/terraformersmc/terraform/wood/api/block/PillarLogHelper.java:
```diff
@@ -12,70 +12,66 @@ private PillarLogHelper() {
 	 */
-	public static PillarBlock ofNether(MapColor color) {
-		return new PillarBlock(AbstractBlock.Settings.create()
+	public static AbstractBlock.Settings settingsNether(MapColor color) {
```
```suggestion
	public static AbstractBlock.Settings createNetherSettings(MapColor color) {
```

haykam821 commented on terraform-wood-api-v1/src/main/java/com/terraformersmc/terraform/wood/api/block/PillarLogHelper.java:
```diff
@@ -12,70 +12,66 @@ private PillarLogHelper() {
 	 */
-	public static PillarBlock of(MapColor wood, MapColor bark) {
-		return new PillarBlock(AbstractBlock.Settings.create()
+	public static AbstractBlock.Settings settings(MapColor wood, MapColor bark) {
```
```suggestion
	public static AbstractBlock.Settings createSettings(MapColor wood, MapColor bark) {
```

haykam821 commented on terraform-wood-api-v1/src/main/java/com/terraformersmc/terraform/wood/api/block/PillarLogHelper.java:
```diff
@@ -12,70 +12,66 @@ private PillarLogHelper() {
 	 */
-	public static PillarBlock of(MapColor color) {
-		return new PillarBlock(AbstractBlock.Settings.create()
+	public static AbstractBlock.Settings settings(MapColor color) {
```
```suggestion
	public static AbstractBlock.Settings createSettings(MapColor color) {
```

gniftygnome committed:
Adjust name of new pillar log helpers.

I was trying to keep them short, but I didn't really like the names either, so I'm fine with this.

Co-authored-by: haykam821 <24855774+haykam821@users.noreply.github.com>

gniftygnome committed:
Update ancillary gradle cruft to match version 8.8.

gniftygnome committed:
Update to 1.21.2-rc1.  Fix wood test.

gniftygnome committed:
Fix the raft entity renderers.

gniftygnome commented:
> * Now that they have a texture, I see the custom rafts do not render quite right.  There is a tiny one texture pixel high space above any edge viewed across the raft where you can see into the water.  I have absolutely no idea what is going on there.

This is fixed now in [1248758](https://github.com/TerraformersMC/Terraform/pull/95/commits/1248758da4677b49e90d207d4b57aec22089d830).

gniftygnome committed:
Add DFU support so old boats can become new boats.

This is the nice clean way I wanted to do but can't because somebody decided the DFU should init before mods do...

gniftygnome committed:
Version which enables mods to register boat IDs early.

This is the icky version I hate to have to make, but the DFU initializes way too early to do this in any nice way...

gniftygnome commented:
OK, I think the API is fine; perhaps things could have better names but we can always fix that in 1.21.4 or whatever.  If we do want to improve this in 1.21.2/3, it would be via API additions or internal changes, which shouldn't be disruptive.  I'll release this as `12.0.0-alpha.1`.
